File: example/src/example/TreePrinter.py
# mypy: disable-error-code="no-redef"

from typing import Optional

import AST
from Utils import addToClass


class TreePrinter:
    @staticmethod
    def safe_print_tree(obj: Optional[AST.Tree], indent_level: int) -> None:
        """Helper function to handle primitives and objects with `print_tree`."""
        prefix = "|  " * indent_level
        if obj is not None:
            obj.print_tree(indent_level)
        else:
            print(f"{prefix}None")

    @addToClass(AST.Assign)
    def print_tree(self: AST.Assign, indent_level=0) -> None:
        print("|  " * indent_level + "=")
        TreePrinter.safe_print_tree(self.var, indent_level + 1)
        TreePrinter.safe_print_tree(self.expr, indent_level + 1)

    @addToClass(AST.If)
    def print_tree(self: AST.If, indent_level=0) -> None:
        print("|  " * indent_level + "IF")
        TreePrinter.safe_print_tree(self.condition, indent_level + 1)
        print("|  " * indent_level + "THEN")
        for stmt in self.then:
            stmt.print_tree(indent_level + 1)
        if self.else_:
            print("|  " * indent_level + "ELSE")
            for stmt in self.else_:
                stmt.print_tree(indent_level + 1)

    @addToClass(AST.While)
    def print_tree(self: AST.While, indent_level=0) -> None:
        print("|  " * indent_level + "WHILE")
        print("|  " * (indent_level + 1) + "CONDITION")
        TreePrinter.safe_print_tree(self.condition, indent_level + 2)
        print("|  " * (indent_level + 1) + "BODY")
        for stmt in self.body:
            stmt.print_tree(indent_level + 2)

    @addToClass(AST.For)
    def print_tree(self: AST.For, indent_level=0) -> None:
        print("|  " * indent_level + "FOR")
        print("|  " * (indent_level + 1) + self.var.name)
        print("|  " * (indent_level + 1) + "RANGE")
        TreePrinter.safe_print_tree(self.range.start, indent_level + 2)
        TreePrinter.safe_print_tree(self.range.end, indent_level + 2)
        print("|  " * (indent_level + 1) + "BODY")
        for stmt in self.body:
            stmt.print_tree(indent_level + 2)

    @addToClass(AST.Apply)
    def print_tree(self: AST.Apply, indent_level=0) -> None:
        assert isinstance(self.ref, AST.SymbolRef)
        print("|  " * indent_level + f"{self.ref.name}")
        print("|  " * (indent_level + 1) + "ARGUMENTS")
        for arg in self.args:
            TreePrinter.safe_print_tree(arg, indent_level + 2)

    @addToClass(AST.Range)
    def print_tree(self: AST.Range, indent_level=0) -> None:
        print("|  " * indent_level + "RANGE")
        TreePrinter.safe_print_tree(self.start, indent_level + 1)
        TreePrinter.safe_print_tree(self.end, indent_level + 1)

    @addToClass(AST.SymbolRef)
    def print_tree(self: AST.SymbolRef, indent_level=0) -> None:
        print("|  " * indent_level + self.name)

    @addToClass(AST.VectorRef)
    def print_tree(self: AST.VectorRef, indent_level=0) -> None:
        print("|  " * indent_level + "VECTORREF")
        print("|  " * (indent_level + 1) + "ARGUMENTS")
        TreePrinter.safe_print_tree(self.vector, indent_level + 2)
        TreePrinter.safe_print_tree(self.element, indent_level + 2)

    @addToClass(AST.MatrixRef)
    def print_tree(self: AST.MatrixRef, indent_level=0) -> None:
        print("|  " * indent_level + "MATRIXREF")
        print("|  " * (indent_level + 1) + "ARGUMENTS")
        TreePrinter.safe_print_tree(self.matrix, indent_level + 2)
        TreePrinter.safe_print_tree(self.row, indent_level + 2)
        TreePrinter.safe_print_tree(self.col, indent_level + 2)

    @addToClass(AST.Literal)
    def print_tree(self: AST.Literal, indent_level=0) -> None:
        print("|  " * indent_level + str(self.value))

    @addToClass(AST.Break)
    def print_tree(self: AST.Break, indent_level=0) -> None:
        print("|  " * indent_level + "BREAK")

    @addToClass(AST.Continue)
    def print_tree(self: AST.Continue, indent_level=0) -> None:
        print("|  " * indent_level + "CONTINUE")

    @addToClass(AST.Return)
    def print_tree(self: AST.Return, indent_level=0) -> None:
        print("|  " * indent_level + "RETURN")
        TreePrinter.safe_print_tree(self.expr, indent_level + 1)

    @staticmethod
    def print_result(result: list[AST.Statement]) -> None:
        for r in result:
            r.print_tree(0)

File: example/src/example/TreePrinter.scala
package example

object TreePrinter extends TreeVisitor[Int]:
  override val handleNull: Int => Unit = _ => ()

  override given Process[AST.Block] = indent =>
    block =>
      println("|  " * indent + "BLOCK")
      block.statements.foreach(_.visit(indent + 1))

  override given Process[AST.Literal] = indent => literal => println("|  " * indent + literal.value.toString)

  override given Process[AST.SymbolRef] = indent => symbolRef => println("|  " * indent + symbolRef.name)

  override given Process[AST.VectorRef] = indent =>
    vectorRef =>
      println("|  " * indent + "VECTORREF")
      println("|  " * (indent + 1) + "ARGUMENTS")
      vectorRef.vector.visit(indent + 2)
      vectorRef.element.visit(indent + 2)

  override given Process[AST.MatrixRef] = indent =>
    matrixRef =>
      println("|  " * indent + "MATRIXREF")
      println("|  " * (indent + 1) + "ARGUMENTS")
      matrixRef.matrix.visit(indent + 2)
      if matrixRef.row.ne(null) then matrixRef.row.visit(indent + 2)
      if matrixRef.col.ne(null) then matrixRef.col.visit(indent + 2)

  override given Process[AST.Apply] = indent =>
    apply =>
      println("|  " * indent + s"${apply.ref.name}")
      println("|  " * (indent + 1) + "ARGUMENTS")
      apply.args.foreach(_.visit(indent + 2))

  override given Process[AST.Range] = indent =>
    range =>
      println("|  " * indent + "RANGE")
      range.start.visit(indent + 1)
      range.end.visit(indent + 1)

  override given Process[AST.Assign] = indent =>
    assign =>
      println("|  " * indent + "=")
      assign.ref.visit(indent + 1)
      assign.expr.visit(indent + 1)

  override given Process[AST.If] = indent =>
    ifStmt =>
      println("|  " * indent + "IF")
      ifStmt.condition.visit(indent + 1)
      println("|  " * indent + "THEN")
      ifStmt.thenBlock.statements.foreach(_.visit(indent + 1))
      if ifStmt.elseBlock.ne(null) then
        println("|  " * indent + "ELSE")
        ifStmt.elseBlock.statements.foreach(_.visit(indent + 1))

  override given Process[AST.While] = indent =>
    whileStmt =>
      println("|  " * indent + "WHILE")
      println("|  " * (indent + 1) + "CONDITION")
      whileStmt.condition.visit(indent + 2)
      println("|  " * (indent + 1) + "BODY")
      whileStmt.body.statements.foreach(_.visit(indent + 2))

  override given Process[AST.For] = indent =>
    forStmt =>
      println("|  " * indent + "FOR")
      println("|  " * (indent + 1) + forStmt.varRef.name)
      println("|  " * (indent + 1) + "RANGE")
      forStmt.range.start.visit(indent + 2)
      forStmt.range.end.visit(indent + 2)
      println("|  " * (indent + 1) + "BODY")
      forStmt.body.statements.foreach(_.visit(indent + 2))

  override given Process[AST.Return] = indent =>
    returnStmt =>
      println("|  " * indent + "RETURN")
      returnStmt.expr.visit(indent + 1)

  override given Process[AST.Continue] = indent => _ => println("|  " * indent + "CONTINUE")

  override given Process[AST.Break] = indent => _ => println("|  " * indent + "BREAK")
